tests/chainermn_tests/functions_tests/test_point_to_point_communication.py

```diff
@@ -9,224 +8,262 @@
+    }, {
+        'dtype': numpy.float32,
+    }]]
+
+function = sigmoid
+evaluation = mse
+
+
+def create_communicator(gpu, param):
+    if gpu:
+        communicator = chainermn.create_communicator('flat')
+        device = communicator.intra_rank
+        chainer.cuda.get_device_from_id(device).use()
+    else:
+        communicator = chainermn.create_communicator('naive')
+        device = -1
+
+    if communicator.size < 2:
+        pytest.skip('This test is for multinode')
+
+    return communicator
+
+
+def create_x(param, communicator):
+    x = chainer.Variable(
+        numpy.arange(10).reshape(1, 10).astype(param.dtype) / 10)
+
+    device = communicator.intra_rank
+
+    if device >= 0:
+        x.to_gpu()
 
-        self.rank_send = (self.communicator.rank + 1) % self.communicator.size
-        self.rank_recv = (self.communicator.rank - 1) % self.communicator.size
+    return x
 
-        # Activation function.
-        self.f = chainer.functions.sigmoid
 
-        # Evaluation function.
-        self.evaluation = chainer.functions.mean_squared_error
+def create_models(param, communicator):
+    model = chainer.links.Linear(
+        10, 10, initialW=_init_w(communicator.rank))
+    entire_model = [chainer.links.Linear(
+        10, 10, initialW=_init_w(l))
+        for l in range(communicator.size)]
 
-        # Input data.
-        self.x = chainer.Variable(
-            numpy.arange(10).reshape(1, 10).astype(numpy.float32) / 10)
+    device = communicator.intra_rank
 
-        self.model = chainer.links.Linear(
-            10, 10, initialW=self._init_w(self.communicator.rank))
-        self.entire_model = [chainer.links.Linear(
-            10, 10, initialW=self._init_w(l))
-            for l in range(self.communicator.size)]
-        self.device = device
+    if device >= 0:
+        model.to_gpu()
+        for model_ in entire_model:
+            model_.to_gpu()
 
-        if device >= 0:
-            self.x.to_gpu()
-            self.model.to_gpu()
-            for model in self.entire_model:
-                model.to_gpu()
+    return (model, entire_model)
 
-    def _init_w(self, l):
-        return 1.0 * numpy.arange(100).reshape(10, 10).astype(numpy.float32) \
-            / ((l + 1) * 100)
 
-    def check_communication(self):
-        if self.communicator.rank == 0:
+def _init_w(l):
+    return 1.0 * numpy.arange(100).reshape(10, 10) \
+        / ((l + 1) * 100)
+
+
+def check_communication(gpu, param):
+    with chainer.using_config('dtype', param.dtype):
+        communicator = create_communicator(gpu, param)
+        rank_send = (communicator.rank + 1) % communicator.size
+        rank_recv = (communicator.rank - 1) % communicator.size
+        x = create_x(param, communicator)
+        (model, entire_model) = create_models(param, communicator)
+
+        if communicator.rank == 0:
             # Input process.
-            y = self.f(self.model(self.x))
+            y = function(model(x))
```

Comment:
Isn't `sigmoid(model(x))` just enough?

Reply:
I want to avoid hard coding. Using 'function' can be more convenient when we want to change to a different one, say ReLU. The same goes to 'evaluation'.